antismash/common/secmet/features/cds_feature.py

```diff
@@ -60,6 +60,7 @@ def __init__(self, location: FeatureLocation, translation: str, locus_tag: str =
             raise TypeError("product must be a string, not %s", type(product))
         self.product = product
         self.transl_table = int(translation_table)
+        self.codon_start = int(codon_start)
```

Comment:
so are you fine to crash when /codon_start not in ["1", "2", "3"]?

Reply:
To raise a ValueError, yes. Otherwise I can provide it -5 and we'll go ahead and potentially spend hours in analysis that will generate garbage output or crash with some hard to debug error on something that seems barely related.

In general, if you know that there is a tightly restricted range on what values are valid, it's easy and safer to ensure that what's provided is in that restricted range.

Reply:
As for the translation_table, i think the setter for codon_start is not in plans, right? Do you mind if, instead of a getter, i will leave this still a public property, just like translation_table? With the ValueError check of course.

Reply:
Python type hints are only hints. They don't provide any kind of runtime security. The reason we use them is because we had quite a lot of contributed code in antiSMASH 4 that took us many hours to debug because we had to work out what was coming in and why it was wrong. With type hints, we can run the type checkers to find any errors with semantic code analysis before it crashes in some poor users code.

The `self.foo = int(foo)` style will catch anything that comes in as a non-integer and either convert it to an integer, e.g. `"1"` or raise an error that it couldn't be converted to an integer, e.g. `"one", [], None` etc. Before my recent changes to the translation table, one of the acceptable (but incorrectly treated) values was `"Standard"`, which caused some problems. So the int conversion there helps catch any cases of that the moment people tried it.

For the codon_start specifically, you don't need the `int()` so much, since you can do `if not 0 <= codon-start <= 2: raise ...` to catch the issue.  The problem main with leaving it a public attribute and not a property is that you can't do that check when it's assigned to. It's not such a big deal, since we aren't expecting any of the analyses to change any properties of a feature in this fashion. Using properties is something we've done in 5 specifically to prevent contributors from just adding things to features as happened in earlier versions, which caused some very strange behaviours.